source/slang/slang-ir-sccp.cpp

```diff
@@ -228,6 +273,460 @@ struct SCCPContext
+                        break;
+                    case kIROp_Int16Type:
+                    case kIROp_UInt16Type:
+                        intVal = intVal & 0xFFFF;
+                        break;
+                    case kIROp_IntType:
+                    case kIROp_UIntType:
+                    case kIROp_BoolType:
+                        intVal = intVal & 0xFFFFFFFF;
+                        break;
+                    default:
+                        break;
+                    }
+                    resultVal = getBuilder()->getIntValue(type, (IRIntegerValue)intVal);
+                }
+                break;
+            default:
+                return LatticeVal::getAny();
+            }
+            break;
+        case kIROp_FloatType:
+        case kIROp_DoubleType:
+        case kIROp_HalfType:
+            switch (irConstant->getOp())
+            {
+            case kIROp_FloatLit:
+                resultVal = getBuilder()->getFloatValue(
+                    type, (IRFloatingPointValue)irConstant->value.floatVal);
+                break;
+            case kIROp_IntLit:
+            case kIROp_BoolLit:
+                resultVal = getBuilder()->getFloatValue(
+                    type, (IRFloatingPointValue)irConstant->value.intVal);
+                break;
+            default:
+                return LatticeVal::getAny();
+            }
+            break;
+        case kIROp_BoolType:
+            switch (irConstant->getOp())
+            {
+            case kIROp_FloatLit:
+                resultVal = getBuilder()->getBoolValue(irConstant->value.floatVal != 0);
+                break;
+            case kIROp_IntLit:
+            case kIROp_BoolLit:
+                {
+                    resultVal = getBuilder()->getBoolValue(irConstant->value.intVal != 0);
+                }
+                break;
+            default:
+                return LatticeVal::getAny();
+            }
+        }
+        if (!resultVal)
+            return LatticeVal::getAny();
+        return LatticeVal::getConstant(resultVal);
+    }
+
+    template<typename TIntFunc, typename TFloatFunc>
+    LatticeVal evalBinaryImpl(
+        IRType* type,
+        LatticeVal v0,
+        LatticeVal v1,
+        const TIntFunc& intFunc,
+        const TFloatFunc& floatFunc)
+    {
+        SLANG_SCCP_RETURN_IF_NONE_OR_ANY(v0)
+        auto c0 = as<IRConstant>(v0.value);
+        SLANG_SCCP_RETURN_IF_NONE_OR_ANY(v1)
+        auto c1 = as<IRConstant>(v1.value);
+        IRInst* resultVal = nullptr;
+        switch (type->getOp())
+        {
+        case kIROp_Int8Type:
+        case kIROp_Int16Type:
+        case kIROp_IntType:
+        case kIROp_Int64Type:
+        case kIROp_UInt8Type:
+        case kIROp_UInt16Type:
+        case kIROp_UIntType:
+        case kIROp_UInt64Type:
+        case kIROp_BoolType:
+            resultVal = getBuilder()->getIntValue(type, intFunc(c0->value.intVal, c1->value.intVal));
```

Comment:
One issue here (not worth addressing right now) is that we don't properly limit the range constant value we store based on the type. We should probably have a function that takes an `IRIntegerLiteralVal` and an `IROp` for an integer type, and limits the value to one representable in that type (and the same for floating-point).

The logic for handling `kIROp_Construct` above already handles most of that, so it could in theory just be pulled out to its own function.